### ISSUE TITLE
Callbacks in different webviews not independent

### ISSUE THREAD

**Issue by charlesroddie** [SKLn-Rad/Xam.Plugin.Webview#21]
If you create two webviews wv1 and wv2,
register two callbacks with the same name "cb" on each, with functions f1 and f2,
and wv2 calls "cb", f1 gets run.

Is this expected?

**Comment by SKLn-Rad** [comment]
At the moment, the dictionary the WebViews use is the same. This was mainly done to prevent the need to teardown and recreate them if a view is swapped out and added again later, for example when used for adsense.

Although this does raise a good point. I think I will add a second dictionary to cover my use case and then change the original to use yours as this probably makes sense. Should I mark this down as an enhancement ?

**Comment by charlesroddie** [comment]
Separate dictionaries would be good functionality to have, together with the current `t.RegisterCallback(s,f)` method. So yes would be an enhancement.

If you need a static dictionary, wouldn't a static method `FormsWebView.RegisterCallback(s,f)` be more appropriate?

**Comment by SKLn-Rad** [comment]
Yeah you read my mind there.
So this would be the API:

S: FormsWebView.RegisterGlobalCallback(str, func) : void
I: FormsWebView.RegisterCallback(str, func) : void

S: FormsWebView.UnregisterAllGlobalCallbacks() : void
S: FormsWebView.UnregisterGlobalCallback(params str) : void
S: FormsWebView.GlobalCallbackIsAvailable(str) : bool

I: FormsWebView.UnregisterAllCallbacks() : void
I: FormsWebView.UnregisterCallback(params str) : void
I: FormsWebView.CallbackIsAvailable(str) : bool

; Assuming registering a callback will replace an existing one of the same key and unregistering one which is not valid will not result in an error being thrown.

Sound good?

**Comment by charlesroddie** [comment]
Great!

**Comment by mwpb** [comment]
Hi and thanks for adding this feature! I've just been trying to convert our old RegisterCallbacks into the new RegisterLocalCallbacks and I'm running into a problem when registering the callbacks for a second time. The following mwe seemed to work under the version 1.4.3 but not under 1.5:

https://github.com/mwpb/MultipleWebViewDebug

In UWP if you press the button twice there is a NullReferenceException. If you need any more information or I am using the new feature incorrectly please let me know!

**Comment by SKLn-Rad** [comment]
Hey man, I made some bug fixes to Android which seemed like your issue there but I wasn't able to reproduce anything in my own code. I did notice with yours though that you're registering the callback on the button click. Could you register it on the constructor instead and then call it using InjectJavascript on the button click?

I updated my sample app code with something similar if you want to have a look 👍

**Comment by mwpb** [comment]
Hi, thanks for looking into this. In our app we create and clear away multiple webviews and so the idea behind the above mwe was to demonstrate how on version 1.5 this caused a NullReferenceException. Putting it all in a button click seemed to be the easiest way to simulate this.

Having said all that, in version 1.5.2 this appears to be fixed on UWP and Droid. Thanks!

**Comment by mwpb** [comment]
Hi, 

Hope all is well. 

Unfortunately this behaviour (NullReferenceException on Windows) has returned in 1.6.0. I have read over your previous answer and certainly would be willing to implement things in a different way, but can't quite figure out how that would work. As I mentioned above the tricky bit is that the user should be able to create new webviews on (for instance) a button press. Then we register callbacks on these new instances - and here is where the NullReferenceException is.

Thanks once again for your work on this!

**Comment by mwpb** [comment]
Additionally I have updated the repo here:

**Comment by SKLn-Rad** [comment]
Hey man, I am a little bit confused trying to run your sample code. Here is what is happening for me.

1) I click the button, which registered the OnContentLoaded event and adds the WebView to the stack
2) The control calls GetH();, returning correctly
3) I click the Google button, which then fires GetH(); on google, causing a Javascript here.

This is with version 1.6. Is this the same as what you're getting as it is not a NRE?
I added some more null checking to 1.6.1 as well as another bug fix for Android, so feel free to give this a go first!

**Comment by mwpb** [comment]
Hi, thanks for looking at my code! Ah yes, my example was broken in two ways then... The NullReferenceException is arranged as follows:

1. Click the button called Initial.
2. Brings up web view displaying the html string.
3. Click the button called Initial again.
4. This should clear the previous webview.
5 ...and then throw a NullReferenceException.

In my main project it seemd to me that the exception happened when the localcallback is bound for the second time.

In version 1.6.1 the same happens.

I have updated the html source in the repo to avoid the confusion above (I've just removed the link to google).

Thanks!

**Comment by charlesroddie** [comment]
I tried 1.7 and I am getting "error CS0103: The name 'n' does not exist in the current context"
Is this message helpful at all?

For example in this code for updating the height of the webview:

The html contains
```
function getH() {
    MathJax.Hub.Queue(function () {
        var height = ...
        commH(height);
    });
```
The .net code:
```
self.RegisterLocalCallback("commH", fun x ->
    self.HeightRequest <- float x )
self.add_OnContentLoaded (fun x ->
    safeInjectjavascript("getH();"))
```
The crash occurs with RegisterLocalCallback, regardless of whether the code is called (still crashes when the OnContentLoaded method is removed).

The code works initially for more than one webview, then fails for later ones.

**Comment by SKLn-Rad** [comment]
Fixed with the changes from charles!
Should be available on 1.8.0
https://www.nuget.org/packages/Xam.Plugin.WebView/1.8.0
